### PR TITLE
[deployment] [feature] Planned maintenance banner

### LINKED ISSUE #7675: ✨ Planned maintenance banner
image

## ✨ Summary

[Visit the Figma file](https://www.figma.com/file/guHeIIh8dqFVCks310Wv0G/Style-library?type=design&node-id=346-15261&mode=design&t=zV2fjMYLZCBEKNth-4) • Contact @substrae with design questions

We're going to have some planned downtime during a server OS upgrade.  Let's add a banner to inform our users.

## 💎 Design notes
- The icon should be vertically centred with the title
- Title should be visually bolded
- Expected time and date should be semantically bolded

## ✅ Acceptance Criteria

- [ ] Can turn on and off at runtime (use runtime flag)
- [ ] Outage time range, date, starttime configurable at runtime as well
- [ ] Exact copy: "Scheduled server maintenance", "Please note that GC Digital Talent will be unavailable for an expected period of 2 hours on Saturday, August 26th, 2023 at (time with timezone). We apologize for any inconvenience."
- [ ] Visible everywhere.

### PR DESCRIPTION
**Pull request by yonikid15**
🤖 Resolves #7675 

## 👋 Introduction

- Adds the new maintenance banner component. The banner sits under the header and spans the entire screen informing users during planned downtime, server upgrades, etc.

## 🕵️ Details

You will need to set the environment variables under `MAINTENANCE_BANNER_X` in your env file, for example:
```
MAINTENANCE_BANNER_PUBLIC_DATE="2023-09-06"
MAINTENANCE_BANNER_START_DATE="2023-09-09 12:00:00"
MAINTENANCE_BANNER_DURATION=1
```

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Open up any page and see no maintenance banner
2. Set all the maintenance banner values in your .env file
3. See maintenance banner at the top of the page (under the GOC logo and language toggle)
4. Ensure banner styling matches the figma design. Ensure text is correct and styling matches.

## 📸 Screenshot

image

## 🚚 Deployment

- environment variable changes for server OS upgrade banner

### CODE REVIEW

File: apps/web/src/components/Layout/MaintenanceBanner.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import React from "react";
+import { useIntl } from "react-intl";
+import isAfter from "date-fns/isAfter";
```
Comment (petertgiles) [inline apps/web/src/components/Layout/MaintenanceBanner.tsx:3]: You've got a linting warning here.  See a similar override in packages/date-helpers/src/index.ts.

File: apps/web/src/components/Layout/MaintenanceBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  const serverMaintenanceDuration = getRuntimeVariable(
+    "MAINTENANCE_BANNER_DURATION",
+  );
+
+  console.log(serverMaintenanceDate);
+  const showMaintenanceBanner =
+    maintenanceBannerPublicDate && serverMaintenanceDate
+      ? (isEqual(
+          parseDateTimeUtc(maintenanceBannerPublicDate),
+          parseDateTimeUtc(currentDate()),
```
Comment (petertgiles) [inline apps/web/src/components/Layout/MaintenanceBanner.tsx:33]: This is a little odd, to turn the current date to an ISO UTC string then parse it back to a Date.  You can just do new Date() to get the current date as a Date object.

File: apps/web/.env.example [hunk 3]
```diff
@@ -15,3 +15,7 @@ FEATURE_SKILL_LIBRARY=true
 LOG_CONSOLE_LEVEL="debug"
 LOG_APPLICATIONINSIGHTS_LEVEL="warning"
 
+# Maintenance banner
```
Comment (petertgiles) [inline apps/web/.env.example:18]: Could you please add the example values (like from the PR) as comments?  For the duration, please note hours as the unit.

File: apps/web/src/components/Layout/MaintenanceBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    <Alert.Root type="warning" live banner>
+      <Alert.Title>
+        {intl.formatMessage({
+          defaultMessage: "Scheduled server maintenance",
+          id: "H0RLNn",
+          description: "Heading for a server maintenance banner.",
+        })}
+      </Alert.Title>
+      <p>
+        {intl.formatMessage(
```
Comment (petertgiles) [inline apps/web/src/components/Layout/MaintenanceBanner.tsx:56]: Oops, watch out for the scenarios where the formatting function has special text.  I'd suggest just using one of the date-fns functions directly
image
image

**Review by petertgiles — CHANGES_REQUESTED** [review]
Nice, this is looking great!  :sunglasses: A few small things to touch up, below.

File: apps/web/src/components/Layout/MaintenanceBanner.tsx (hunk 2 above)
Comment (yonikid15) [reply apps/web/src/components/Layout/MaintenanceBanner.tsx]: Ya good point 😆 Fixed by [88f4c06](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7832/commits/88f4c06c3867f6804a4f87b8a770c5ce0b996bfd)

File: apps/web/src/components/Layout/MaintenanceBanner.tsx (hunk 1 above)
Comment (yonikid15) [reply apps/web/src/components/Layout/MaintenanceBanner.tsx:4]: Perfect thanks, fixed by [88f4c06](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7832/commits/88f4c06c3867f6804a4f87b8a770c5ce0b996bfd)

File: apps/web/.env.example (hunk 3 above)
Comment (yonikid15) [reply apps/web/.env.example:18]: Nice catch 👍 Fixed by [5a203ec](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7832/commits/5a203eca2ec6a1a9413f2720371b60a70da05198)

File: apps/web/src/components/Layout/MaintenanceBanner.tsx (hunk 4 above)
Comment (yonikid15) [reply apps/web/src/components/Layout/MaintenanceBanner.tsx:60]: Nice catch again, should be fixed by [88f4c06](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7832/commits/88f4c06c3867f6804a4f87b8a770c5ce0b996bfd)

Should I leave in the possibility of decimal numbers? or just round up using `Math.ceil()`

Comment (petertgiles) [reply apps/web/src/components/Layout/MaintenanceBanner.tsx:60]: Leaving as a text is probably fine.  We could put in integers, decimals, or even things like "1-2".

File: apps/web/src/components/Layout/MaintenanceBanner.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        intl,
+      })
+    : null;
+
+  const duration =
+    serverMaintenanceDuration && !!Math.abs(+serverMaintenanceDuration)
+      ? Math.abs(+serverMaintenanceDuration)
+      : null;
+
+  const showMaintenanceBanner =
```
Comment (petertgiles) [inline apps/web/src/components/Layout/MaintenanceBanner.tsx:38]: I think there's a bug here.  I'm before the public date but it's still showing.
Also, the message disappears when the current date is the same as the public date.  Maybe it's backwards?
image

File: apps/web/src/components/Layout/MaintenanceBanner.tsx (hunk 4 above)
Comment (petertgiles) [reply apps/web/src/components/Layout/MaintenanceBanner.tsx:60]: Oh, it only works for numbers.  I think that's fine. :+1:

File: apps/web/src/components/Layout/MaintenanceBanner.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    maintenanceBannerPublicDate && serverMaintenanceDate
+      ? (isEqual(parseDateTimeUtc(maintenanceBannerPublicDate), new Date()) ||
+          isAfter(parseDateTimeUtc(maintenanceBannerPublicDate), new Date())) &&
+        isBefore(
```
Comment (petertgiles) [inline apps/web/src/components/Layout/MaintenanceBanner.tsx:42]: Did you mean to compare the two environment variables here?

**Review by petertgiles — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Layout/MaintenanceBanner.tsx (hunk 5 above)
Comment (yonikid15) [reply apps/web/src/components/Layout/MaintenanceBanner.tsx:38]: Ya the values should be reversed 😓 Fixed by [d85c93e](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7832/commits/d85c93ea09b93ddc6d8cc92a0702a10051a6455e)

File: apps/web/src/components/Layout/MaintenanceBanner.tsx (hunk 6 above)
Comment (yonikid15) [reply apps/web/src/components/Layout/MaintenanceBanner.tsx:42]: I think they just needed to be reversed right. 

If current date is equal or after the public date time then show the banner.

**Review by petertgiles — DISMISSED** [review]
Nice, it's working well.  :+1:

**Comment by yonikid15** [comment]
These are some style fixes Josh mentioned:

- [ ] Icon alignment is a bit weird
- [ ] There seems to be a weird margin below the banner
- [ ] I don't think we need the shadow :thinking_face:

image

**Review by petertgiles — APPROVED** [review]
Noice.  :+1:

**Pull request merged**